src/net_processing.cpp

```diff
@@ -6270,6 +6270,10 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
 
                 // Determine transactions to relay
                 if (fSendTrickle) {
+                    if (m_txreconciliation && m_txreconciliation->IsPeerRegistered(pto->GetId())) {
+                        // Make transactions added to the reconciliation set during the last interval available
+                        m_txreconciliation->ReadyDelayedTransactions(pto->GetId());
```

Comment:
20772c077832592265bd6a2876aa6b4cb7dbde7d

with this code, and especially the 5/2 trickle delay (very long), previous performance measurements become outdated.

My idea was to [delay the responses](https://github.com/bitcoin/bitcoin/pull/21515/commits/f99fa469115856aefa9799b3fc62ba09985933ad). And by this time, the hope was flooding does enough job so that scanning the sets through many connections does not work.

Also, adding to a set already happens on a trickle... so that's another level of protection

Reply:
Just talked to Bruno a bit more about this.

I think it's fair to compare this delayed set mechanism to what i have in the code there.
I like my solution more because:
- its synced across peers (handles better spying across multiple conns i think?)
- simpler code
- more flexibility (doesn't have to be same trickling as adding to sets or fanouting)

Interested in what other people think though.

Reply:
I think the current approach mimics what you were doing with your old one. I haven't picked up the trickle reduction commit not the delayed response, but I was planning to in the next PR.

Your approach used to add transactions to the reconciliation set during INV building, which happens on trickle intervals. So data was added to `to_be_announced` (delayed) on `RelayTransaction` and then made available on trickle. The current approach adds data to the delayed set on `RelayTransaction` and makes it available on trickle too.

So I don't think the previous performance measurements become outdated, building the whole Erlay PR on this should yield similar results.

Reply:
I agree now, it's equivalent. Still wrapping my head around `RelayTransaction` refactoring :)

---------------------------

A difference could be in handling dependencies: during the trickle window after the first transaction arrives.

Say, a parent arrives, and then a child arrives in a second. In my approach we may notice the child when handling (announcement-wise) the parent (because handling the parent is postponed via trickle).
In your approach, you handle the parent independently.

Is there any use of knowing about the child while handling (announcement-wise) the parent? I don't think so. You probably could save some efforts of adding the parent to reconset, and then removing it, but that should be cheap anyway.

-------------------------

(the child-before-parent case is entirely different, as the child-to-arrive-early doesn't normally hit the announcement step while in orphanage)

Reply:
Found an answer! ([here](https://gist.github.com/naumenkogs/514ff3901960161a7b7ee08449840768)) 

```
09:42:53<@gmaxwell> Existing relay logic checks that transactions are still in the mempool before relaying them. I think the issue there would go away if instead of keeping around some erlay datastructure you just keep growing the queue of the peers transactions to relay until its time to reconcile, enh?  then the existing logic to check if things are still in the mempool is sufficient.
```

Reply:
I'm planning to simulate both cases and report back

Reply:
I simulated this to see if there was any difference between doing the decision-making before or after trickle.

I didn't do it for dependant transactions, just for a single one, trying to answer whether choosing fanout peers at scheduling time or at relay time made any difference. The answer is no, given the number of peers that know about the transaction at both times is almost the same. 

The approach, simulation and results can be found here: https://srgi.notion.site/Select-fanout-candidates-at-relay-time-instead-of-at-relay-scheduling-time-1617b3fef97780b9aa9bc52ec6fb4049

It is worth mentioning that this applies to fully deployed erlay, which I think is what we should be designing for.